llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -5108,6 +5108,50 @@ Instruction* cs6475_optimizer_brensen(Instruction *I) {
   // END BRENSEN VILLEGAS
 }
 
+// BEGIN DOMINIC KENNEDY
+uint64_t get_max_cmp(uint64_t add, uint64_t or_i) {
+  add -= 1;
+  uint64_t msb = 32 - __builtin_clz(add & or_i);
+  if (msb == 0) {
+    return 1;
+  }
+
+  uint64_t mask = (1UL << (msb - 1)) - 1;
+  return 1 + (mask & add) + (mask & or_i);
+}
+
+uint64_t extract_rounded_int(ConstantInt *x) {
+  uint64_t mask = (1UL << x->getUniqueInteger().getBitWidth()) - 1;
```

Comment:
so I'll merge your patch but you should know that this code is incorrect. LLVM APInts can have arbitrary width and in fact we often see widths >64 while compiling real codes. But here you're working in a 64-bit integer so this code will fail badly for the >64 case.